### ISSUE TITLE
Download / Parse / Load bioRxiv and medRxiv articles

### ISSUE THREAD

**Issue by jankrepl** [BlueBrain/Search#459]
## 🚀 Feature
* Check licencing
* Investigate whether it is possible to bulk download papers from Arxiv
* Implement a parser for these papers

## Motivation
More articles

## Pitch
More articles is good

## Alternatives
Maybe investigate similar sources like bioRxiv etc.

**Comment by pafonta** [comment]
## Licence

_[...] bulk TDM, which authors explicitly consent to during submission to bioRxiv [...].  The TDM repository is **not intended as a source for further redistribution** of articles posted on bioRxiv, or their derivatives, **nor does it grant others permission to re-host content** posted on bioRxiv.  For most articles submitted to bioRxiv, authors retain copyright and reuse rights. **If you build indexing services or tools based on the full text of articles, you must therefore link back to the text hosted at bioRxiv rather than re-host content.** For reuse/redistribution of individual articles or their derivatives, please consult the licensing terms applied by the authors, which are provided in the metadata. In most cases, this will require you to contact the copyright holder in advance to obtain permission._
([source](https://www.biorxiv.org/tdm))

## Download

_Full-text access via Amazon S3 is available via a **requester pays** bucket. [...] The bioRxiv AWS bucket is located in the US East (N. Virginia) [...]._
([source](https://www.biorxiv.org/tdm))

➡️  Bulk download is charged:
[Using Requester Pays buckets for storage transfers and usage](https://docs.aws.amazon.com/AmazonS3/latest/userguide/RequesterPaysBuckets.html)

Here is how to bulk download articles:

1. Create an `AWS` account.
2. Get an `IAM` role which will receive the bills.
3. Install AWS `CLI` v2:
  [Installing, updating, and uninstalling the AWS CLI version 2](https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2.html)
4. Configure the CLI and assume the given IAM role:
  TL;DR [Switching to an IAM role (AWS CLI)](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_use_switch-role-cli.html) (See _Example scenario: Switch to a production role_)
  [Configuration basics](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-quickstart.html)
  [Configuration and credential file settings](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html)
  [Using an IAM role in the AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-role.html)
5. Check everything works fine and summary statistics with:
  `aws s3 ls s3://biorxiv-src-monthly  --request-payer requester --summarize --human-readable`
  [List S3 objects and common prefixes under a prefix or all S3 buckets](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/s3/ls.html)
6. Bulk download articles:
  `aws s3 cp s3://biorxiv-src-monthly bioarxiv/ --request-payer requester --recursive`:
  [Copy a local file or S3 object to another location locally or in S3](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/s3/cp.html)

## Parsing

TODO

## Loading

TODO

**Comment by pafonta** [comment]
Regarding `arXiv` (`arXiv/q-bio`):

There is `no` XML version of the full text of these articles. The source files are in `TeX/LaTeX` or in `PDF`. ([source](https://arxiv.org/help/bulk_data_s3))

➡️  We might want to `report` the download / parsing / loading of the papers from `arXiv/q-bio`.

Indeed, processing these papers requires new components and sets new challenges.

**Comment by FrancescoCasalegno** [comment]
> Regarding `arXiv` (`arXiv/q-bio`):
> 
> There is `no` XML version of the full text of these articles. The source files are in `TeX/LaTeX` or in `PDF`. ([source](https://arxiv.org/help/bulk_data_s3))

Just for the sake of completeness, this seems to be true in most cases. But in principle, there may also be some articles on `arXiv` that have other source files in other formats than `TeX`, including `HTML` (source 1) (source 2).

That being said, so far I have only seen `arXiv` articles with `PDF` available, some also with `TeX`, but none with `HTML`.

**Comment by pafonta** [comment]
### Proposed action for loading

Have `bbs_database parse`:
1. recursively look for files to parse in the `input_path`,
2. look only for the file format (e.g. XML, JSON) associated with the `input_type`.

(details: section 'loading' in https://github.com/BlueBrain/Search/issues/459#issuecomment-931355275)

Please confirm you have read this proposal and made your comments below if needed:

- [ ] @FrancescoCasalegno 
- [ ] @jankrepl 
- [ ] @EmilieDel 
- [ ] @Stannislav

**Comment by FrancescoCasalegno** [comment]
Hi @pafonta,

Yes I think your proposal of recursive parsing makes perfect sense. As for

> file format (e.g. XML, JSON) associated with the `input_type`

I think we should just be a bit careful to cover all possible cases. E.g. for `JATS XML`, are we sure that `.xml` is the only possible extension? Didn't we see some `.nxml` or something like that in the `PMC-OA` papers?
(Maybe @EmilieDel knows this better.)

--- 

Maybe we could have an optional parameter (whose the defaults could be the extensions we expect to be associated to the `input_type`) `--extension` for `bbs_database parse` so that we could pass something like `--extension=.xml,.nxml` if we know the extension of the files at runtime.
This is just an idea though, feel free to move forward with the implementation :)

**Comment by FrancescoCasalegno** [comment]
## Summary of Discussion

**Option 1** – `bbs_dataset parse` called once per file.
_Drawbacks_: It would imply significant overhead, see https://github.com/BlueBrain/Search/pull/443#issue-997039708

**Option 2** – pipe results of a `find` into `bbs_dataset parse`. 
_Drawbacks_: It would mean passing a string containing 1M files as an argument (it may break). Moreover this would mean we have to change the logic of the functionalities of `bbs_database`.

**Option 3** – Support an argument `--match-filename` (or better name) to identify which files inside a given directory will be matched.
_Drawbacks_: It may not be clear what comes out of the regex. (Optional): We can have a `--dry-run` option to print out which files would be matched.

## Conclusion

Let's go with Option 3 ! 💪

**Comment by pafonta** [comment]
All elements have been done (see [here](https://github.com/BlueBrain/Search/issues/459#issuecomment-931355275)). Closing.
